src/Tools/dotnet-dsrouter/DiagnosticsServerRouterCommands.cs

```diff
@@ -0,0 +1,142 @@
+using System;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Diagnostics.Tools.DSRouter
+{
+    public class DiagnosticsServerRouterLauncher : DiagnosticsServerRouterRunner.Callbacks
+    {
+        public CancellationToken CommandToken { get; set; }
+        public bool SuspendProcess { get; set; }
+        public bool Verbose { get; set; }
+
+        public void OnRouterStarted(string boundTcpServerAddress)
+        {
+            if (ProcessLauncher.Launcher.HasChildProc)
+            {
+                string diagnosticPorts = boundTcpServerAddress + (SuspendProcess ? ",suspend" : ",nosuspend");
+                if (ProcessLauncher.Launcher.ChildProc.StartInfo.Arguments.Contains("${DOTNET_DiagnosticPorts}", StringComparison.OrdinalIgnoreCase))
+                {
+                    ProcessLauncher.Launcher.ChildProc.StartInfo.Arguments = ProcessLauncher.Launcher.ChildProc.StartInfo.Arguments.Replace("${DOTNET_DiagnosticPorts}", diagnosticPorts);
+                    diagnosticPorts = "";
+                }
+
+                ProcessLauncher.Launcher.Start(diagnosticPorts, CommandToken, Verbose, Verbose);
+            }
+        }
+
+        public void OnRouterStopped()
+        {
+            ProcessLauncher.Launcher.Cleanup();
+        }
+    }
+
+    public class DiagnosticsServerRouterCommands
+    {
+        public static DiagnosticsServerRouterLauncher Launcher { get; } = new DiagnosticsServerRouterLauncher();
+
+        public DiagnosticsServerRouterCommands()
+        {
+        }
+
+        public async Task<int> RunIpcClientTcpServerRouter(CancellationToken token, string ipcClient, string tcpServer, int runtimeTimeout, bool verbose)
+        {
+            checkLoopbackOnly(tcpServer);
+
+            using CancellationTokenSource cancelRouterTask = new CancellationTokenSource();
+            using CancellationTokenSource linkedCancelToken = CancellationTokenSource.CreateLinkedTokenSource(token, cancelRouterTask.Token);
+
+            using var factory = new LoggerFactory();
+            factory.AddConsole(verbose ? LogLevel.Debug : LogLevel.Information, false);
+
+            Launcher.SuspendProcess = true;
+            Launcher.Verbose = verbose;
+            Launcher.CommandToken = token;
+
+            var routerTask = DiagnosticsServerRouterRunner.runIpcClientTcpServerRouter(linkedCancelToken.Token, ipcClient, tcpServer, runtimeTimeout == Timeout.Infinite ? runtimeTimeout : runtimeTimeout * 1000, factory.CreateLogger("dotnet-dsrounter"), Launcher);
+
+            while (!linkedCancelToken.IsCancellationRequested)
+            {
+                await Task.WhenAny(routerTask, Task.Delay(250)).ConfigureAwait(false);
+                if (routerTask.IsCompleted)
+                    break;
+
+                if (Console.KeyAvailable)
+                {
+                    ConsoleKey cmd = Console.ReadKey(true).Key;
```

Comment:
I believe this will throw under output redirection, e.g., `cmd > file.log`. In other tools we check if there is redirection and skip key checks for the status loop.